src/common/hook-setting.js

```diff
@@ -13,7 +13,24 @@ options.hook((data) => {
   });
 });
 
-export default function hook(key, update) {
+/**
+ * When an option is updated elsewhere (or when a yet unresolved options.ready will be fired),
+ * calls the specified `update` function or assigns the specified `prop` in `update` object.
+ * Also, when `prop` is specified, option.get() is called explicitly right away,
+ * but only if options.ready is resolved or a `handle` function is specified.
+ * @param {string} key - option name
+ * @param {function(value) | Object} update - either a function or the target object
+ * @param {Object} [config]
+ * @param {string} config.prop - property name to be assigned in the target
+ * @param {function(value)} [config.handle] - transformer function to call on the value
+ * @return {function}
+ */
+export default function hook(key, update, { prop, handle } = {}) {
```

Comment:
It seems that `update` only appears as an object. Maybe we don't need to support it as a function any more.
ps. I think the relationship between `update` and `prop` is a little hard to understand. Though it's not a big deal, I prefer the original style.

Reply:
> It seems that `update` only appears as an object.

vm-sync.vue uses a function but I can convert it to use the declarative syntax too

> ps. I think the relationship between `update` and `prop` is a little hard to understand. 

Do you mean the code inside the function is hard to understand or the usage?

> Though it's not a big deal, I prefer the original style.

You mean you like this:
```js
hookSetting('filters.searchScope', (value) => {
  filters.searchScope = value;
});
hookSetting('filters.showEnabledFirst', (value) => {
  filters.showEnabledFirst = value;
});
hookSetting('filters.sort', (value) => {
  filters.sort.set(value);
});
options.ready.then(() => {
  filters.searchScope = options.get('filters.searchScope');
  filters.sort.set(options.get('filters.sort'));
  filters.showEnabledFirst = options.get('filters.showEnabledFirst');
});
```
more than this:
```js
Object.keys(filters).forEach(name => {
  hookSetting(`filters.${name}`, filters, { prop: name });
});
```
right?

Reply:
What about flattening it like `hookSetting(key, targetObj, propName, transformValueFunc)`?

Reply:
Changing `update` to `targetObj` looks good to me.

> ps. I think the relationship between update and prop is a little hard to understand.

I mean if `update` is a function, `prop` does not make sense. If `update` should be kept any way, I prefer `hookSetting(key, update | { target, prop, transform })`.

Reply:
Ah, yeah, I've considered that too, but then went the implicit/confusing route. Okay let's do `hookSetting(key, update | { target, prop, transform })`

Reply:
Pushed the new version.